tests/extraction_tests/test_round_three_extraction.py

```diff
@@ -338,7 +407,49 @@ def test_extract_risk(mock_risk_sheet, mock_project_lookup, mock_programme_looku
     assert project_xor_programme.all()
 
 
-# TODO: Add test of whole extract, and run some assertions ie that projects line up as expected between tabs etc.
+def test_full_ingest(mock_ingest_full_extract):
+    """
+    Tests on top-level ingest function for Towns Fund Round 3.
+
+    Test entity relationships etc.
+    """
+
+    # test that sheets that map 1:1 with projects:sheet rows have the same unique project id's
+    valid_projects_for_extract = set(mock_ingest_full_extract["Project Details"]["Project ID"])
+    assert set(mock_ingest_full_extract["Project Progress"]["Project ID"]) == valid_projects_for_extract
+    assert set(mock_ingest_full_extract["Funding Comments"]["Project ID"]) == valid_projects_for_extract
+    assert set(mock_ingest_full_extract["Funding"]["Project ID"]) == valid_projects_for_extract
+    assert set(mock_ingest_full_extract["Private Investments"]["Project ID"]) == valid_projects_for_extract
+
+    # test tables of mixed programme/project only have valid project id's (including NaN rows mapped only to programme)
+    valid_projects_for_extract.add(np.nan)
+    assert not set(mock_ingest_full_extract["Output_Data"]["Project ID"]) - valid_projects_for_extract
+    assert not set(mock_ingest_full_extract["Outcome_Data"]["Project ID"]) - valid_projects_for_extract
+    assert not set(mock_ingest_full_extract["RiskRegister"]["Project ID"]) - valid_projects_for_extract
+
+    # test only valid programmes for this extract are in programme-level tables
+    valid_programmes_for_extract = {mock_ingest_full_extract["Place Details"]["Programme ID"][0], np.nan}
+    assert not set(mock_ingest_full_extract["Programme_Ref"]["Programme ID"]) - valid_programmes_for_extract
+    assert not set(mock_ingest_full_extract["Project Details"]["Programme ID"]) - valid_programmes_for_extract
+    assert not set(mock_ingest_full_extract["Programme Progress"]["Programme ID"]) - valid_programmes_for_extract
+    assert not set(mock_ingest_full_extract["Funding Questions"]["Programme ID"]) - valid_programmes_for_extract
+    assert not set(mock_ingest_full_extract["Outcome_Data"]["Programme ID"]) - valid_programmes_for_extract
+    assert not set(mock_ingest_full_extract["RiskRegister"]["Programme ID"]) - valid_programmes_for_extract
```

Comment:
Are any of these cases valid data ingest cases? 

- "If any project sheets have nan project id values then this test will fail": It should be impossible for this to happen, unless the spreadsheet is broken - in which case the pipeline will fail. I could alter the test data to replicate a "broken" spreadsheet, and in the test assert that it raises whatever error it causes (most likely some random key error) - but the options for this are infinite :/ 

- "if any proj/prog sheets have both ids as nans": This will fail at DB as DB condition does not allow it. It should only happen in reality if someone enters an invalid project name (overrides the dropdown). In this case we should only really test how we handle the error, which we currently don't do in the pipeline. I could add the test case, but it wouldn't prove anything other than doing this is allowed by the pipeline (but caught later).

- "programme sheets have nans for programme id" fair point, I'll split this slightly different case out

Reply:
also, if "any proj/prog sheets have both ids as nans" - this is tested elsewhere in test_extract_outcomes

The tests in this PR are supposed to only be in relation to the high-level relational integrity

Reply:
yeah you're right - i've been thinking all day yesterday about broken spreadsheets... these tests obviously assume a valid spreadsheet